File: CS/ViewModel.cs
// Developer Express Code Central Example:
// How to parallelize data-intensive operations on in-memory data in Instant Feedback UI mode
// 
// In this demo, the DevExpress Grid Control for WPF operates in Instant Feedback
// data binding mode on in-memory data. All operations on data (e.g. sorting,
// groping, filtering, summary calculation, etc.) are performed asynchronously and
// parallelized on multiple processors. This allows the computing power of your
// hardware to be utilized to the full extent without UI freezing.
// See also:
// WPF
// Data Grid Control – PLINQ Data Support
// (http://community.devexpress.com/blogs/theprogressbar/archive/2011/08/22/wpf-data-grid-control-plinq-data-support.aspx)
// Binding
// to In-Memory Data Using PLINQ
// (http://documentation.devexpress.com/#WPF/CustomDocument10472)
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E3382

using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;

namespace PLinqDataSource {
    public class PLinqViewModel{

        public PLinqViewModel() {
            ListSource = new OrderDataListSource(1000000);
        }

        public virtual OrderDataListSource ListSource { get; set; }
    }

    public class OrderDataListSource : IListSource {
        List<OrderData> orders;
        int count = 0;
        static object Locker = new object();

        public OrderDataListSource(int count) {
            this.count = count;
        }

        void GenerateOrders() {
            orders = new List<OrderData>(count);
            Random rnd = new Random();
            int customersCount = Data.Customers.Count;
            int productsCount = Data.Products.Count;
            for(int i = 0; i < count; i++) {
                OrderData orderData = new OrderData();
                orderData.OrderId = i + 1;
                orderData.CustomerName = Data.Customers[rnd.Next(customersCount)];
                orderData.OrderDate = DateTime.Today.Subtract(TimeSpan.FromDays(rnd.Next(180)));
                KeyValuePair<string, decimal> product = Data.Products[rnd.Next(productsCount)];
                orderData.ProductName = product.Key;
                orderData.Price = product.Value;
                orderData.Quantity = rnd.Next(200) + 1;
                orders.Add(orderData);
            }
        }
        public IList GetList() {
            if(orders == null) {
                lock(Locker) {
                    if(orders == null) {
                        GenerateOrders();
                    }
                }
            }
            return orders;
        }
        public bool ContainsListCollection {
            get { return false; }
        }
    }

    public class OrderData {
        public int OrderId { get; set; }
        public string CustomerName { get; set; }
        public DateTime OrderDate { get; set; }
        public string ProductName { get; set; }
        public decimal Price { get; set; }
        public int Quantity { get; set; }
    }
}

File: VB/ViewModel.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel

Namespace PLinqDataSource

    Public Class PLinqViewModel
        Implements INotifyPropertyChanged

        Private orderDataListSource As OrderDataListSource

        Public Sub New()
            ListSource = New OrderDataListSource(1000000)
        End Sub

        Public Property ListSource As OrderDataListSource
            Get
                Return orderDataListSource
            End Get

            Set(ByVal value As OrderDataListSource)
                If orderDataListSource Is value Then Return
                orderDataListSource = value
                OnPropertyChanged("ListSource")
            End Set
        End Property

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Overridable Sub OnPropertyChanged(ByVal propertyName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class

    Public Class OrderDataListSource
        Implements IListSource

        Private orders As List(Of OrderData)

        Private count As Integer = 0

        Private Shared Locker As Object = New Object()

        Public Sub New(ByVal count As Integer)
            Me.count = count
        End Sub

        Private Sub GenerateOrders()
            orders = New List(Of OrderData)(count)
            Dim rnd As Random = New Random()
            Dim customersCount As Integer = Customers.Count
            Dim productsCount As Integer = Products.Count
            For i As Integer = 0 To count - 1
                Dim orderData As OrderData = New OrderData()
                orderData.OrderId = i + 1
                orderData.CustomerName = Customers(rnd.Next(customersCount))
                orderData.OrderDate = Date.Today.Subtract(TimeSpan.FromDays(rnd.Next(180)))
                Dim product As KeyValuePair(Of String, Decimal) = Products(rnd.Next(productsCount))
                orderData.ProductName = product.Key
                orderData.Price = product.Value
                orderData.Quantity = rnd.Next(200) + 1
                orders.Add(orderData)
            Next
        End Sub

        Public Function GetList() As IList Implements IListSource.GetList
            If orders Is Nothing Then
                SyncLock Locker
                    If orders Is Nothing Then
                        GenerateOrders()
                    End If

                End SyncLock
            End If

            Return orders
        End Function

        Public ReadOnly Property ContainsListCollection As Boolean Implements IListSource.ContainsListCollection
            Get
                Return False
            End Get
        End Property
    End Class

    Public Class OrderData

        Public Property OrderId As Integer

        Public Property CustomerName As String

        Public Property OrderDate As Date

        Public Property ProductName As String

        Public Property Price As Decimal

        Public Property Quantity As Integer
    End Class
End Namespace
